### PR TITLE
Dns resolution vpc

### PR DESCRIPTION
**Pull request by ujjwal-ibm**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: ibm/service/vpc/resource_ibm_is_vpc.go [hunk 1]
```diff
@@ -644,6 +734,47 @@ func vpcGet(d *schema.ResourceData, meta interface{}, id string) error {
...
+		}
+		if vpc.Dns.Resolver != nil {
+			switch reflect.TypeOf(vpc.Dns.Resolver).String() {
+			case "*vpcv1.VpcdnsResolverTypeDelegated":
+				{
+					resolver := vpc.Dns.Resolver.(*vpcv1.VpcdnsResolverTypeDelegated)
+					currentDnsResolver := map[string]interface{}{}
+					currentDnsResolver[isVPCDnsResolverType] = resolver.Type
+
+					currentDnsResolver[isVPCDnsResolverVpc] = setDnsResolverVpc(resolver.VPC)
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_vpc.go:754]: isVPCDnsResolverVpc should be defined in the schema

**Review by deepaksibm — CHANGES_REQUESTED** [review]
Also add patch support for dns

File: ibm/service/vpc/resource_ibm_is_vpc.go [hunk 2]
```diff
@@ -1170,3 +1389,290 @@ func suppressNullAddPrefix(k, old, new string, d *schema.ResourceData) bool {
...
+		VPCModel, err := resourceIBMIsVPCMapToVpcdnsResolverVPCPatch(modelMap["vpc"].([]interface{})[0].(map[string]interface{}))
+		if err != nil {
+			return model, err
+		}
+		model.VPC = VPCModel
+	}
+	return model, nil
+}
+
+func resourceIBMIsVPCMapToVpcdnsResolverVPCPatch(modelMap map[string]interface{}) (vpcv1.VpcdnsResolverVPCPatchIntf, error) {
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_vpc.go:1484]: "id" and "CRN" should be mutually exclusive

File: ibm/service/vpc/resource_ibm_is_vpc.go [hunk 3]
```diff
@@ -112,7 +134,173 @@ func ResourceIBMISVPC() *schema.Resource {
...
+										Computed:    true,
+										Description: "The type of the DNS resolver used for the VPC.- `delegated`: DNS server addresses are provided by the DNS resolver of the VPC               specified in `dns.resolver.vpc`.- `manual`: DNS server addresses are specified in `dns.resolver.manual_servers`.- `system`: DNS server addresses are provided by the system.",
+									},
+									isVPCDnsResolverVpc: &schema.Schema{
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "The VPC whose DNS resolver provides the DNS server addresses for this VPC.The VPC may be remote and therefore may not be directly retrievable.",
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												isVPCDnsResolverVpcCrn: &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_vpc.go:196]: isVPCDnsResolverVpcCrn and isVPCDnsResolverVpcId. should be marked Optional as well
and should be mutually exclusive.

File: ibm/service/vpc/resource_ibm_is_vpc.go [hunk 4]
```diff
@@ -1170,3 +1389,290 @@ func suppressNullAddPrefix(k, old, new string, d *schema.ResourceData) bool {
...
+				return model, err
+			}
+			manualServers = append(manualServers, *manualServersItemModel)
+		}
+		model.ManualServers = manualServers
+	}
+	if modelMap["type"] != nil && modelMap["type"].(string) != "" {
+		model.Type = core.StringPtr(modelMap["type"].(string))
+	}
+	if modelMap["vpc"] != nil && len(modelMap["vpc"].([]interface{})) > 0 {
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_vpc.go:1474]: Handle this scenario 
"Specify null to remove an existing VPC."
https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/swagger-ui.html?feature=is-vpc-dns-sharing-topology-for-endpoint-gateways#/VPCs/update_vpc

File: website/docs/r/is_vpc.html.markdown [hunk 5]
```diff
@@ -53,6 +71,34 @@ Review the argument references that you can specify for your resource.
 - `default_network_acl_name` - (Optional, String) Enter the name of the default network access control list (ACL).
 - `default_security_group_name` - (Optional, String) Enter the name of the default security group.
 - `default_routing_table_name` - (Optional, String) Enter the name of the default routing table.
+- `dns` - (Optional, List) The DNS configuration for this VPC.
+  
+  Nested scheme for `dns`:
+  - `enable_hub` - (Optional, Boolean) Indicates whether this VPC is enabled as a DNS name resolution hub.
+  - `resolver` - (Optional, List) The zone list this backup policy plan will create snapshot clones in.
+    Nested scheme for `resolver`:
+      - `manual_servers` - (Optional, Integer) The DNS servers to use for this VPC, replacing any existing servers. All the DNS servers must either: **have a unique zone_affinity**, or **not have a zone_affinity**.
```
Comment (deepaksibm) [inline website/docs/r/is_vpc.html.markdown:80]: Should be (Optional, List) of address and zone_affinity

File: website/docs/r/is_vpc.html.markdown [hunk 6]
```diff
@@ -53,6 +71,34 @@ Review the argument references that you can specify for your resource.
...
+        ~> **Note:** 
+              Updating from `manual` requires dns resolver `manual_servers` to be specified as null.<br/>
+              Updating to `manual` requires dns resolver `manual_servers` to be specified and not empty.<br/>
+              Updating from `delegated` requires `dns.resolver.vpc` to be specified as null.
+      - `vpc` - (Optional, String) The VPC to provide DNS server addresses for this VPC. The specified VPC must be configured with a DNS Services custom resolver and must be in one of this VPC's DNS resolution bindings.
+
+        ~> **Note:** 
+          Specify null to remove an existing VPC.<br/>
+          This property must be set if and only if dns resolver type is `delegated`.
+
```
Comment (deepaksibm) [inline website/docs/r/is_vpc.html.markdown:101]: Also add to ## Attribute reference section

Comment (ujjwal-ibm) [reply website/docs/r/is_vpc.html.markdown:101]: addressed

File: website/docs/r/is_vpc.html.markdown [hunk 7]
```diff
@@ -53,6 +71,34 @@ Review the argument references that you can specify for your resource.
...
+        ~> **Note:** 
+          `delegated`: DNS server addresses will be provided by the resolver for the VPC specified in dns.resolver.vpc. Requires dns.enable_hub to be false.<br/>
+          `manual`: DNS server addresses are specified in `manual_servers`.<br/>
+          `system`: DNS server addresses will be provided by the system and depend on the configuration.
+
+        ~> **Note:** 
+              Updating from `manual` requires dns resolver `manual_servers` to be specified as null.<br/>
+              Updating to `manual` requires dns resolver `manual_servers` to be specified and not empty.<br/>
+              Updating from `delegated` requires `dns.resolver.vpc` to be specified as null.
+      - `vpc` - (Optional, String) The VPC to provide DNS server addresses for this VPC. The specified VPC must be configured with a DNS Services custom resolver and must be in one of this VPC's DNS resolution bindings.
```
[the remark this answers is not in the document]
Comment (ujjwal-ibm) [reply website/docs/r/is_vpc.html.markdown]: addressed

File: ibm/service/vpc/resource_ibm_is_vpc.go (hunk 4 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_vpc.go:1481]: addressed

File: ibm/service/vpc/resource_ibm_is_vpc.go (hunk 3 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_vpc.go:197]: addressed

File: ibm/service/vpc/resource_ibm_is_vpc.go (hunk 2 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_vpc.go:1491]: addressed

File: ibm/service/vpc/resource_ibm_is_vpc.go (hunk 1 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_vpc.go]: addressed

**Pull request closed without merging**